srml/staking/src/slashing.rs

```diff
@@ -0,0 +1,772 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A slashing implementation for NPoS systems.
+//!
+//! For the purposes of the economic model, it is easiest to think of each validator
+//! of a nominator which nominates only its own identity.
+//!
+//! The act of nomination signals intent to unify economic identity with the validator - to take part in the
+//! rewards of a job well done, and to take part in the punishment of a job done badly.
+//!
+//! There are 3 main difficulties to account for with slashing in NPoS:
+//!   - A nominator can nominate multiple validators and be slashed via any of them.
+//!   - Until slashed, stake is reused from era to era. Nominating with N coins for E eras in a row
+//!     does not mean you have N*E coins to be slashed - you've only ever had N.
+//!   - Slashable offences can be found after the fact and out of order.
+//!
+//! The algorithm implemented in this module tries to balance these 3 difficulties.
+//!
+//! First, we only slash participants for the _maximum_ slash they receive in some time period,
+//! rather than the sum. This ensures a protection from overslashing.
+//!
+//! Second, we do not want the time period (or "span") that the maximum is computed
+//! over to last indefinitely. That would allow participants to begin acting with
+//! impunity after some point, fearing no further repercussions. For that reason, we
+//! automatically "chill" validators and withdraw a nominator's nomination after a slashing event,
+//! requiring them to re-enlist voluntarily (acknowledging the slash) and begin a new
+//! slashing span.
+//!
+//! Typically, you will have a single slashing event per slashing span. Only in the case
+//! where a validator releases many misbehaviors at once, or goes "back in time" to misbehave in
+//! eras that have already passed, would you encounter situations where a slashing span
+//! has multiple misbehaviors. However, accounting for such cases is necessary
+//! to deter a class of "rage-quit" attacks.
+//!
+//! Based on research at https://research.web3.foundation/en/latest/polkadot/slashing/npos/
+
+use super::{
+	EraIndex, Trait, Module, Store, BalanceOf, Exposure, Perbill, SessionInterface,
+	NegativeImbalanceOf,
+};
+use sr_primitives::traits::{Zero, Saturating};
+use support::{
+	StorageValue, StorageMap, StorageDoubleMap,
+	traits::{Currency, OnUnbalanced, Imbalance},
+};
+use rstd::vec::Vec;
+use codec::{Encode, Decode};
+
+/// The proportion of the slashing reward to be paid out on the first slashing detection.
+/// This is f_1 in the paper.
+const REWARD_F1: Perbill = Perbill::from_percent(50);
+
+/// The index of a slashing span - unique to each stash.
+pub(crate) type SpanIndex = u32;
```

Comment:
Do we really need `pub(crate)` everywhere? The module is not really public anyway.

Reply:
what do you suggest instead? `pub(crate)` is better than `pub` for this case, and we need items to be used in the module above.

Reply:
If the module is not public, there isn't really a difference, is there? I prefer `pub` as it's more concise.

Reply:
There is a difference in intent. It also stops `pub use` from higher up to fail, meaning the author of that line would have to stop and think about changing the intent of the code.